src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+	uint32_t m_pen_previous_input;
+	uint32_t m_pen_target;
+	uint32_t m_pages_previous_input;
+	uint32_t m_effective_page;
+
+	uint16_t m_rtc[0x100/2]{};
+
+	std::unique_ptr<uint32_t[]> m_midi_regs;
+	uint32_t m_midi_busy_count;
+
+	required_ioport m_io_cpu_config;
+	required_ioport m_io_page_config;
+	required_ioport m_io_video_config;
+
+private:
+	uint8_t m_memcache[0x800]{};
+	enum memcache_seq : uint32_t
+	{
+		BITS_EMPTY = 0,
+		BITS_0     = 0x0000ff00,
+		BITS_0_1   = 0x00ff0001, // Command 0xa0 store bit 0
+								 // Command 0xa1 move to next bit?
+		BITS_0_1_7 = 0xff000107, // Reset
+		BITS_6     = 0x0000ff06, // Query (1 if ready for next events)
+		BITS_6_7   = 0x00ff0607, // Query (0 if events parsed successfully)
+								 // Command 0xa0 store bit 1
+								 // Command 0xa1 load bit
+		BITS_7     = 0x0000ff07,
+		BITS_7_1   = 0x00ff0701, // Start command
+		BITS_ERR   = 0xffffffff,
+	};
+	uint32_t m_memcache_seq;
+	enum memcache_state : uint8_t
+	{
+		IDLE = 0,
+		READ_COMMAND,
+		READ_ADDRESS_BYTE,
+		READ_ACK,
+		READ_RAM_BYTE,
+		WRITE_RAM_BYTE,
+		CONSUME_UNTIL_IDLE,
+	};
+	uint32_t m_memcache_addr;
+	uint8_t m_memcache_data;
+	uint8_t m_memcache_i;
+	uint8_t m_memcache_state;
+	std::queue<uint8_t> m_memcache_state_after;
+	std::queue<uint8_t> m_memcache_events;
 };
 
-void sega_beena_state::beena_arm7_map(address_map &map)
+/*
+    FIXME: All of these have guessed timing.
+
+    Animation durations in hardware captures suggest that an IRQ is raised every VBLANK.
+    Palette changes can occur for each scanline via FIQ callback set by games.
+
+    These variables in BIOS RAM may be worth following:
+    - 0x20003ee0: FIQ enabled;
+    - 0x20003ee1: IRQ enabled;
+    Along with pairs used by interrupt callbacks:
+    - 0x20003ed1..0x20003ed4: status;
+    - 0x20003ee2..0x20003ee5: 1 if busy?
+*/
+TIMER_DEVICE_CALLBACK_MEMBER(sega_9h0_0008_state::scanline)
 {
-	map(0x00000000, 0x000001ff).rom().bankr("cartbank");
-	map(0x80000000, 0x807fffff).rom().bankr("cartbank");
+	irq_wait_speedup();
+
+	// Pixel clock approximation derived from palette lookup tables applied per scanline,
+	// used for gradient backgrounds in "Partner In TV".
+	//
+	// FIXME:
+	// - Rainbow arc in "Cars 2" before bridge section is not rendered;
+	// - Foreground drop animation in "Cooking Beena" title screen is skipped;
+	m_scanline = param / m_scale;
+	m_video_regs[0x4/4] = 0x10000 * (20 + m_scanline);
+
+	uint8_t irq_frequency = m_io_cpu_config->read() & 0xf;
+	if ((m_scanline % (SCREEN_H / irq_frequency)) == 0) {
+		uint8_t video_status = (m_video_regs[0] & 0x10) ^ 0x10;
+		m_video_regs[0] = (m_video_regs[0] & 0xffffffef) | video_status;
+
+		update_rtc();
+
+		request_irq();
+
+		// FIXME: Needed for .au data processing in "Kazoku Minna no Nouryoku Trainer"
+		m_requested_fiq = true;
+	}
+
+	request_fiq();
 }
 
+static const uint32_t irq_wait_signature[] = {
+   0xe3a03001,
+   0xe5c03000,
+   0xe5d03000,
+   0xe3530000,
+   0x1afffffc
+};
```

Comment:
Please declare this inside the function that uses it, and use `SCREAMING_SNAKE_CASE` since it behaves like a constant.